examples/tutorials/core_api/3_hpsearch.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+
+"""
+Stage 3: Let's add hyperparameter search to our model.
+"""
+
+import logging
+import pathlib
+import sys
+import time
+
+import determined as det
+
+
+def save_state(x, latest_batch, trial_id, checkpoint_directory):
+    with checkpoint_directory.joinpath("state").open("w") as f:
+        f.write(f"{x},{latest_batch},{trial_id}")
+
+
+def load_state(trial_id, checkpoint_directory):
+    checkpoint_directory = pathlib.Path(checkpoint_directory)
+    with checkpoint_directory.joinpath("state").open("r") as f:
+        x, latest_batch, ckpt_trial_id = [int(field) for field in f.read().split(",")]
+    if ckpt_trial_id == trial_id:
+        return x, latest_batch
+    else:
+        return x, 0
+
+
+def main(core_context, latest_checkpoint, trial_id, increment_by):
+    x = 0
+
+    starting_batch = 0
+    if latest_checkpoint is not None:
+        with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+            x, starting_batch = load_state(trial_id, path)
+
+    # NEW: iterate through the core_context.searcher.operations() to decide how long to train for.
+    batch = starting_batch
+    last_checkpoint_batch = None
+    for op in core_context.searcher.operations():
+        # NEW: we will use a while loop for easier accounting of absolute lengths.
+        while batch < op.length:
+            x += increment_by
+            time.sleep(.1)
+            print("x is now", x)
+            batch += 1
+            if batch % 10 == 9:
```

Comment:
this is going to checkpoint at slightly wrong times.  What do you think of switching to `while batch < op.length` in all the examples to avoid counting inconsistencies between examples?  Originally I avoided that because the `for i in range(n)` is so idiomatic in the early examples.

Reply:
in examples 1 and 2 it means `while batch < 100`